### PR TITLE
Use safetensors in llama

### PR DESCRIPTION
**Pull request by python273**
I'll rebase when #710 and #711 merged

### CODE REVIEW

**Comment by geohot** [comment]
So why do I want this?

**Comment by python273** [comment]
seems mainly useful for saving concated weights, also loads a little bit faster on cpu (i added hack for numpy). If you think it's not needed in the example, feel free to close

7B:
pickled 4.5s
safetensors 3.6s

13B:
pickled + concat 8.9s + 1m36s
safetensors 7.1s

File: examples/llama.py [hunk 1]
```diff
@@ -216,59 +219,100 @@ def sample(logits, temperature):
...
+      with safe_open(filename, framework="numpy") as f:
+        for k in (t := tqdm(f.keys())):
+          p = get_child(model, k)
+          v = f.get_tensor(k)
+          if is_cpu:
+            vt = Tensor.empty(*p.shape, dtype=dtypes.from_np(v))
+            vt.lazydata.realized = CPUBuffer(v)
+            p.assign(vt).realize()
+          else:
+            p.assign(v).realize()
```
Comment (python273) [inline examples/llama.py:297]: ^ hack for numpy

**Comment by geohot** [comment]
Re: 13B, I assume you save and load, right? You can do that without safetensors.

But more importantly, why is that concat so slow! It should be one copy.

**Comment by python273** [comment]
> Re: 13B, I assume you save and load, right? You can do that without safetensors.

yep, saving first. You mean can be saved with pickle?

> I expect pickled + concat to be around 15s.

yep, if I manually concat, it's 9s unpickle + 6s concat 🤔 

```python
        if w0.shape[0] != mv.shape[0]:
          # mv.assign(w0.cat(w1, dim=0))
          ax = 0
        elif w0.shape[1] != mv.shape[1]:
          # mv.assign(w0.cat(w1, dim=1))
          ax = 1
        else: raise RuntimeError("what axis mismatch?")

        print(mv.shape, w0.shape, w1.shape, ax)

        np0 = w0.lazydata.realized._buf
        np1 = w1.lazydata.realized._buf
        vt = Tensor.empty(*mv.shape, dtype=w0.dtype)
        vt.lazydata.realized = CPUBuffer(np.concatenate((np0, np1), axis=ax))
        mv.assign(vt).realize()
```

**Comment by geohot** [comment]
Now that's an interesting bug to fix!

**Comment by python273** [comment]
it's implemented as pads and add. is there a way to fix without adding a special op? 🤔 
Screenshot 2023-03-19 at 05 31 51

**Comment by geohot** [comment]
What device are you using? On GPU etc..., that should be free.

**Comment by python273** [comment]
it was `CPU=1`

`CLANG=1` is 51s to concat

**Pull request closed without merging**